### DISCUSSION TITLE
contributing/development/configuring_an_ide/visual_studio

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#96] [User-contributed notes]
# contributing/development/configuring_an_ide/visual_studio

Visual Studio Community is a Windows-only IDE by Microsoft that's free for individual use or non-commercial use within organizations. It has many useful features, such as memory view, performance v...

https://docs.godotengine.org/en/stable/contributing/development/configuring_an_ide/visual_studio.html

**Comment by tiagoacarvalho** [comment]
Using the scons command provided, scons will **not** build. It will only generate the solution.
If you want it to build it directly, use:
```sh
scons platform=windows vsproj=yes dev_build=yes vsproj_gen_only=no
```

**Comment by tiagoacarvalho** [comment]
I believe to have figured it out, how to debug both project and editor tools at once, using Visual Studio at least. If someone knows the best place to put the below information, please do post it there. It took me a good 6 weeks of searching, asking ChatGPT, and throwing wrenches at Windows Processes to stumble upon this.

   1. Build godot from source. I don't know if it's necessary, but I haven't tested without it. Always better to have those debug symbols anyways. I used the following shell script to build it. Be aware: It will remove your current godot installation in the github folder in favor of version 4.3-stable

```shell
cd C:\Users\<user>\Documents\GitHub

Remove-Item -Path C:\Users\<user>\Documents\GitHub\godot -Force -Recurse

git clone --single-branch --branch 4.3-stable 

cd godot

scons platform=windows dev_build=yes vsproj=yes module_mono_enabled=yes debug_symbols=yes
scons p=windows target=template_debug module_mono_enabled=yes
scons p=windows target=template_release module_mono_enabled=yes

cd bin

.\godot.windows.editor.dev.x86_64.mono --headless --generate-mono-glue modules/mono/glue

Move-Item -Path C:\Users\<user>\Documents\GitHub\godot\bin\modules -Destination C:\Users\<user>\Documents\GitHub\godot\ -Force

cd ..

py ./modules/mono/build_scripts/build_assemblies.py --godot-output-dir=./bin --godot-platform=windowshttps://github.com/godotengine/godot.git
```

   2. Open Visual Studio Community with just about everything even loosely related to .NET installed. (I know this is not specific, but I don't exactly know what is and isn't necessary here. I followed the godot documentation's advice when installing Visual Studio up to the in-editor configuration step. Which I just threw a wrench into.
  
  3. Open Godot, enable C# stuffs and build your project with C#, so you generate a Project Solution for it. And check the Exclude Addon Warnings option, just in case. 
  4. Click on the down arrow next to the run with debug options and click in {Project Name} Debug Properties. Configure it as such:
       Executable:  C:\Users\<user>\Documents\GitHub\godot\bin\godot.windows.editor.dev.x86_64.mono.exe
       Command line arguments: --path C:\Users\<user>\Documents\GitHub\<project>--verbose
       Enable native code debugging: yes
This is the configuration that allows you to debug your project.
Now breakpoints and all the such should be working in VS. For the tool scripts:
   1. With godot open and editing the project, go to Debug > Attach to Process
   2. Attach Visual Studio to the Godot Editor process (search 'godot')
And that's it. For as long as the process remains attached, breakpoints will work just fine with tool scripts.

For debugging editor plugins, it's a little simpler, godot docs tell it to you straight. To debug the editor plugin, open another VS instance, with the godot.sln project. Then run and compile it there. You'll get breakpoints in godot's own code and whatever plugin code is added into it.

**Comment by CastonYoung** [comment]
'scons' is not recognized as an internal or external command,
operable program or batch file.

I can't build anything.

**Reply by CastonYoung** [reply]
What exactly is the "Godot root folder"?

**Reply by CastonYoung** [reply]
Okay, so it appears I first need to download SCons...

**Reply by tiagoacarvalho** [reply]
> What exactly is the "Godot root folder"?

As far as I know, it is in whichever folder you clone the repo into. A.k.a the top level folder. The one containing the REAEME.md, as is for any git repo.

I dont recall the exactly where the installer defaults to.

**Comment by Marcos-Belmont** [comment]
ChatGPT Solution:

Okay! Let's take this step by step. If scons isn't working, it's probably because:

SCons isn't installed on your system.

Or the scons command isn't in your PATH.

In the meantime, here's how to install SCons on each system:

* Linux (Ubuntu/Debian):
[bash]
--------------
sudo apt update
sudo apt install scons
--------------

* Arch Linux:
[bash]
--------------
sudo pacman -S scons
--------------

* Windows:
1. Download and install Python (if you don't have it already).
2. Install SCons via pip:
[cmd]
--------------
pip install scons
--------------
Tip: Make sure Python and Python Scripts/ are in your PATH.

* macOS:
With Homebrew:
[bash]
--------------
brew install scons
--------------

** Check if it is working
After installation, type:
[bash]
--------------
scons --version
--------------
If the version appears, everything is fine.

**Reply by LucCADORET** [reply]
`pip install scons` is the preffered method, I'm not even sure the others options work, don't blindly copy-paste an LLM answer, especially when a simple google search gives a more clear result. Of course, python needs to be installed first.

**Comment by BrentAnthuenisDeveloper** [comment]
I get the error: "scons: *** No SConstruct file found." what do you mean with : "the Godot root folder": Is this the root folder of my project or the folder with the godot editor?

**Reply by LucCADORET** [reply]
Since the goal is to compile the engine, then you can guess that the root folder is at the root of the engine files, not the root of your personal project. You should see a file named **SConstruct** (no extention)  there

**Reply by BrentAnthuenisDeveloper** [reply]
The problem was that i was downloading godot from the website rather than from github.
